src/dataqueue/queue.h

```diff
@@ -0,0 +1,311 @@
+    // must fail with an error when a variance is detected.
+    virtual bool isIdempotent() const = 0;
+  };
+
+  // Creates an idempotent DataQueue with a pre-established collection
+  // of entries. All of the entries must also be idempotent otherwise
+  // an empty std::unique_ptr will be returned.
+  static std::shared_ptr<DataQueue> CreateIdempotent(
+      std::vector<std::unique_ptr<Entry>> list);
+
+  // Creates a non-idempotent DataQueue. This kind of queue can be
+  // mutated and updated such that multiple reads are not guaranteed
+  // to produce the same result. The entries added can be of any type.
+  static std::shared_ptr<DataQueue> Create(
+      v8::Maybe<size_t> capped = v8::Nothing<size_t>());
+
+  // Creates an idempotent Entry from a v8::ArrayBufferView. To help
+  // ensure idempotency, the underlying ArrayBuffer is detached from
+  // the BackingStore. It is the callers responsibility to ensure that
+  // the BackingStore is not otherwise modified through any other
+  // means. If the ArrayBuffer is not detachable, nullptr will be
+  // returned.
+  static std::unique_ptr<Entry> CreateInMemoryEntryFromView(
+      v8::Local<v8::ArrayBufferView> view);
+
+  // Creates an idempotent Entry from a v8::BackingStore. It is the
+  // callers responsibility to ensure that the BackingStore is not
+  // otherwise modified through any other means. If the ArrayBuffer
+  // is not detachable, nullptr will be returned.
+  static std::unique_ptr<Entry> CreateInMemoryEntryFromBackingStore(
+      std::shared_ptr<v8::BackingStore> store,
+      size_t offset,
+      size_t length);
+
+  static std::unique_ptr<Entry> CreateDataQueueEntry(
+      std::shared_ptr<DataQueue> data_queue);
+
+  static std::unique_ptr<Entry> CreateStreamBaseEntry(
+      StreamBase* stream_base,
+      BaseObjectPtr<BaseObject> strong_ref);
+
+  static std::unique_ptr<Entry> CreateStreamEntry(
+      Environment* env,
+      v8::Local<v8::Object> obj);
+
+  static std::unique_ptr<Entry> CreateFdEntry(
+      BaseObjectPtr<fs::FileHandle> handle);
+
+  // Creates a Reader that wraps a Reader, making it possible to perform
+  // cross-worker thread reads.
+  static std::unique_ptr<CrossThreadReader> CreateCrossThreadReader(
+      Environment* env,
+      std::unique_ptr<Reader> reader);
+
+  // Creates a Reader for the given queue. If the queue is idempotent,
+  // any number of readers can be created, all of which are guaranteed
+  // to provide the same data. Otherwise, only a single reader is
+  // permitted.
+  virtual std::unique_ptr<Reader> getReader() = 0;
+
+  // Append a single new entry to the queue. Appending is only allowed
+  // when isIdempotent() is false. v8::Nothing<bool>() will be returned
+  // if isIdempotent() is true. v8::Just(false) will be returned if the
+  // data queue is not idempotent but the entry otherwise cannot be added.
+  virtual v8::Maybe<bool> append(std::unique_ptr<Entry> entry) = 0;
```

Comment:
Can you expand a bit on how this method would work?

Reply:
It simply adds the entry to the end of the internal vector if it is allowed to do so, incrementing the total size if the total size is known. The included test has examples.